README.md
<!--lint disable no-literal-urls-->
<h1 align="center">
  SRecord.js
</h1>

A **Highly Experimental and evolving** JavaScript based library to read full S19 strings and create them from an input buffer. The codebase is still under development and is provided without any guarantee, so feel free to contribute to help make it better.

# Table of contents

- [Release Types](#release-types)
  - [Download](#download)
- [Installation](#installation)
- [Security](#security)
- [Contributing](#contributing-to-srecordjs)
- [Support](#support)
- [License](#license)

## Usage Instructions

### Read an S19 string (i.e. read by node-fs)

```js

  import fs from 'fs-extra';
  const file = await fs.readFile(`/path/to/my-s19-file.s19`, 'utf8);

  const response: ReadS19Response = readS19(file, file.length);

  console.log(response.success ? 'Successful!' : `Unsuccesful - ${response.message}`);
  console.log(response.payload?.sRecords);

```

## Convert a buffer to S19 string

```js
const inputContent = "Hello, this is a test of the writeS19 method. This buffer will be converted into an S19 file";
const inputDataBuffer = Buffer.alloc(128);
inputDataBuffer.write(inputContent, "utf8");

const response = writeS19(inputDataBuffer, inputDataBuffer.length, 48, false); //  48 is the length I specified for the data in each S1-3 record (must be between 1-255)

console.log(response.success ? "Successful!" : `Unsuccesful - ${response.message}`);
console.log("S-Record Data:", response.payload);
```

## Release types

- **latest**: Under active development. Code for the latest release is in the
  branch for its major version number.
- **alpha**: Releases that are used solely to test upcoming features and changes. These cannot be guaranteed to be compatible with current releases

Current and LTS releases follow [Semantic Versioning](https://semver.org).

### Download

See [Releases](https://github.com/rsispal/srecord-js/releases) section

## Installation

### Using in your project

#### Via GitHub Package Registry

Ensure you authorise NPM to read from the GitHub Package Registry. The easiest way is to `npm login --registry=npm.pkg.github.com`, but you can also create an `.npmrc` file with the following contents:

```
@rsispal:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken=YOUR_GITHUB_PERSONAL_ACCESS_TOKEN_HERE
```

#### Directly installing Git repository

You can also just directly install the Git repository

Run `npm install --save https://github.com/rsispal/srecord-js`

### From Source

- Clone the project
- Run `npm install`
- Run `npm run build` (production)
- Run `npm run build:dev` (non-production)

## Contributing to SRecord.js

Create a PR if there's any changes and features you can contribute.

## Support

Looking for help? Create an issue and i'll get to you as soon as I can.

## License

[LICENSE](https://github.com/rsispal/srecord-js/blob/master/LICENSE.md) for the full
license text.


CHANGELOG.md
# 1.0.0 (2021-06-29)


### Features

* Initial commit ([22bf660](https://github.com/rsispal/srecord-js/commit/22bf6602562d5251ab13f2cda7f0a566b2d484f6))


src/crc/crc.test.ts
import { LogLevel } from "../types";
import { calculateChecksum } from "./crc";

describe("CRC", () => {
  describe("CRC :: calculateChecksum", () => {
    it('Should calculate a known checksum from a known buffer', () => {
        const buffer = Buffer.from([0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0xFF])
        const response = calculateChecksum(buffer, LogLevel.NONE);
        expect(response).toEqual(136) // 136 XOR 255 & 255
    });

     it('Should return 255 when an empty buffer is provided', () => {
        const buffer = Buffer.from([])
        const response = calculateChecksum(buffer, LogLevel.NONE);
        expect(response).toEqual(255) // 0 XOR 255 & 255
    });

    it('Should return 0 when an buffer with single valule of 0xff is provided', () => {
        const buffer = Buffer.from([0xff])
        const response = calculateChecksum(buffer, LogLevel.NONE);
        expect(response).toEqual(0) // 0 XOR 255 & 255
    });});
});

src/read/read.test.ts
import { readS19 } from "./read";
import { FullSRecordMarker } from "../types";

describe('Read', () => {
    describe('Read :: readS19', () => {
        it ('Should parse inputString1', () => {
             const inputString1 = `S10C000001020304040507F004E5
            S10710F0FFCCFF78B6
            S9030000FC
            `;
            const response = readS19(inputString1, inputString1.length);

            expect(response.success).toBeTruthy();
            expect(response.payload.totalFullSRecords).toEqual(3);
            expect(response.payload.totalDataSize).toEqual(13);

            expect(response.payload.sRecords[0].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE);
            expect(response.payload.sRecords[1].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE);
            expect(response.payload.sRecords[2].sRecordType).toEqual(FullSRecordMarker.S9_RECORD_TYPE);

            expect(response.payload.sRecords[0].address).toEqual(0x00);
            expect(response.payload.sRecords[1].address).toEqual(0x10F0);
            expect(response.payload.sRecords[2].address).toEqual(0x00);

            expect(response.payload.sRecords[0].dataLength).toEqual(0x0C - 3);
            expect(response.payload.sRecords[1].dataLength).toEqual(0x07 - 3);
            expect(response.payload.sRecords[2].dataLength).toEqual(0x03 - 3);

            expect(response.payload.sRecords[0].data.equals(Buffer.from([0x01, 0x02, 0x03, 0x04, 0x04, 0x05, 0x07, 0xF0, 0x04]))).toBeTruthy()
            expect(response.payload.sRecords[1].data.equals(Buffer.from([ 0xFF, 0xCC, 0xFF, 0x78]))).toBeTruthy()
            expect(response.payload.sRecords[2].data.equals(Buffer.from([]))).toBeTruthy()

            expect(response.payload.sRecords[0].checksum).toEqual(0xE5);
            expect(response.payload.sRecords[1].checksum).toEqual(0xB6);
            expect(response.payload.sRecords[2].checksum).toEqual(0xFC);
        });

        it ('Should parse inputString2', () => {

            const inputString2 = `S00F000068656C6C6F202020202000003C
            S11F00007C0802A6900100049421FFF07C6C1B787C8C23783C6000003863000026
            S11F001C4BFFFFE5398000007D83637880010014382100107C0803A64E800020E9
            S111003848656C6C6F20776F726C642E0A0042
            S5030003F9
            S9030000FC`

            const response = readS19(inputString2, inputString2.length);
            expect(response.success).toBeTruthy();
            expect(response.payload.totalFullSRecords).toEqual(6);
            expect(response.payload.totalDataSize).toEqual(82);
      

            expect(response.payload.sRecords[0].sRecordType).toEqual(FullSRecordMarker.S0_RECORD_TYPE)
            expect(response.payload.sRecords[1].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE)
            expect(response.payload.sRecords[2].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE)
            expect(response.payload.sRecords[3].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE)
            expect(response.payload.sRecords[4].sRecordType).toEqual(FullSRecordMarker.S5_RECORD_TYPE)
            expect(response.payload.sRecords[5].sRecordType).toEqual(FullSRecordMarker.S9_RECORD_TYPE)


            expect(response.payload.sRecords[0].address).toEqual(0x0000);
            expect(response.payload.sRecords[1].address).toEqual(0x0000);
            expect(response.payload.sRecords[2].address).toEqual(0x001C);
            expect(response.payload.sRecords[3].address).toEqual(0x0038);
            expect(response.payload.sRecords[4].address).toEqual(0x0003);
            expect(response.payload.sRecords[5].address).toEqual(0x0000);

            expect(response.payload.sRecords[0].dataLength).toEqual(0x0F - 3);
            expect(response.payload.sRecords[1].dataLength).toEqual(0x1F - 3);
            expect(response.payload.sRecords[2].dataLength).toEqual(0x1F - 3);
            expect(response.payload.sRecords[3].dataLength).toEqual(0x11 - 3);
            expect(response.payload.sRecords[4].dataLength).toEqual(0x03 - 3);
            expect(response.payload.sRecords[5].dataLength).toEqual(0x03 - 3);

            expect(response.payload.sRecords[0].data.equals(Buffer.from([0x68, 0x65, 0x6C, 0x6C, 0x6F, 0x20, 0x20, 0x20, 0x20, 0x20, 0x00, 0x00]))).toBeTruthy()
            expect(response.payload.sRecords[1].data.equals(Buffer.from([0x7C, 0x08, 0x02, 0xA6, 0x90, 0x01, 0x00, 0x04, 0x94, 0x21, 0xFF, 0xF0, 0x7C, 0x6C, 0x1B, 0x78, 0x7C, 0x8C, 0x23, 0x78, 0x3C, 0x60, 0x00, 0x00, 0x38, 0x63, 0x00, 0x00]))).toBeTruthy()
            expect(response.payload.sRecords[2].data.equals(Buffer.from([0x4B, 0xFF, 0xFF, 0xE5, 0x39, 0x80, 0x00, 0x00, 0x7D, 0x83, 0x63, 0x78, 0x80, 0x01, 0x00, 0x14, 0x38, 0x21, 0x00, 0x10, 0x7C, 0x08, 0x03, 0xA6, 0x4E, 0x80, 0x00, 0x20]))).toBeTruthy()
            expect(response.payload.sRecords[4].data.equals(Buffer.from([]))).toBeTruthy()
            expect(response.payload.sRecords[5].data.equals(Buffer.from([]))).toBeTruthy()


            expect(response.payload.sRecords[0].checksum).toEqual(0x3C);
            expect(response.payload.sRecords[1].checksum).toEqual(0x26);
            expect(response.payload.sRecords[2].checksum).toEqual(0xE9);
            expect(response.payload.sRecords[3].checksum).toEqual(0x42);
            expect(response.payload.sRecords[4].checksum).toEqual(0xF9);
            expect(response.payload.sRecords[5].checksum).toEqual(0xFC);
        });

    });

})

src/write/write.test.ts
import { writeS19 } from "./write";
import { readS19 } from "../read/read";
import { FullSRecordMarker, LogLevel } from "../types";

describe("Write", () => {
  describe("Write :: writeS19", () => {
    const inputContent = "Hello, this is a test of the writeS19 method. This buffer will be converted into an S19 file";
    const inputDataBuffer = Buffer.alloc(128);
    inputDataBuffer.write(inputContent, "utf8");

    const expectedOutput = `S0030000FC\r\nS133000048656C6C6F2C207468697320697320612074657374206F6620746865207772697465533139206D6574686F642E2054686E\r\nS13300306973206275666665722077696C6C20626520636F6E76657274656420696E746F20616E205331392066696C650000000017\r\nS9030000FC`;

    it("Should convert the inputDataBuffer to an S19 string", () => {
      const response = writeS19(inputDataBuffer, inputDataBuffer.length, 48, false);

      expect(response.success).toBeTruthy();
      expect(response.payload).toEqual(expectedOutput);
    });

    it("Should convert the inputDataBuffer to an S19 string and then I expect to be able to parse it back to this string", () => {
      const response = writeS19(inputDataBuffer, inputDataBuffer.length, 48, false);

      expect(response.success).toBeTruthy();
      expect(response.payload).toEqual(expectedOutput);

      const readResponse = readS19(response.payload, response.payload.length, LogLevel.NONE);

      expect(readResponse.success).toBeTruthy();
      expect(readResponse.payload.totalFullSRecords).toEqual(4);
      expect(readResponse.payload.totalDataSize).toEqual(96);

      expect(readResponse.payload.sRecords[0].sRecordType).toEqual(FullSRecordMarker.S0_RECORD_TYPE);
      expect(readResponse.payload.sRecords[1].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE);
      expect(readResponse.payload.sRecords[2].sRecordType).toEqual(FullSRecordMarker.S1_RECORD_TYPE);
      expect(readResponse.payload.sRecords[3].sRecordType).toEqual(FullSRecordMarker.S9_RECORD_TYPE);

      expect(readResponse.payload.sRecords[0].address).toEqual(0x00);
      expect(readResponse.payload.sRecords[1].address).toEqual(0x0000);
      expect(readResponse.payload.sRecords[2].address).toEqual(0x0030);
      expect(readResponse.payload.sRecords[3].address).toEqual(0x0000);

      expect(readResponse.payload.sRecords[0].dataLength).toEqual(0x03 - 3);
      expect(readResponse.payload.sRecords[1].dataLength).toEqual(0x33 - 3);
      expect(readResponse.payload.sRecords[2].dataLength).toEqual(0x33 - 3);
      expect(readResponse.payload.sRecords[3].dataLength).toEqual(0x03 - 3);

      expect(readResponse.payload.sRecords[0].data.equals(Buffer.from([]))).toBeTruthy();
      expect(
        readResponse.payload.sRecords[1].data.equals(
          Buffer.from([
            0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x2c, 0x20, 0x74, 0x68, 0x69, 0x73, 0x20, 0x69, 0x73, 0x20, 0x61, 0x20, 0x74, 0x65,
            0x73, 0x74, 0x20, 0x6f, 0x66, 0x20, 0x74, 0x68, 0x65, 0x20, 0x77, 0x72, 0x69, 0x74, 0x65, 0x53, 0x31, 0x39, 0x20,
            0x6d, 0x65, 0x74, 0x68, 0x6f, 0x64, 0x2e, 0x20, 0x54, 0x68,
          ])
        )
      ).toBeTruthy();
      expect(
        readResponse.payload.sRecords[2].data.equals(
          Buffer.from([
            0x69, 0x73, 0x20, 0x62, 0x75, 0x66, 0x66, 0x65, 0x72, 0x20, 0x77, 0x69, 0x6c, 0x6c, 0x20, 0x62, 0x65, 0x20, 0x63,
            0x6f, 0x6e, 0x76, 0x65, 0x72, 0x74, 0x65, 0x64, 0x20, 0x69, 0x6e, 0x74, 0x6f, 0x20, 0x61, 0x6e, 0x20, 0x53, 0x31,
            0x39, 0x20, 0x66, 0x69, 0x6c, 0x65, 0x00, 0x00, 0x00, 0x00,
          ])
        )
      ).toBeTruthy();
      expect(readResponse.payload.sRecords[3].data.equals(Buffer.from([]))).toBeTruthy();

      expect(readResponse.payload.sRecords[0].checksum).toEqual(0xfc);
      expect(readResponse.payload.sRecords[1].checksum).toEqual(0x6e);
      expect(readResponse.payload.sRecords[2].checksum).toEqual(0x17);
      expect(readResponse.payload.sRecords[3].checksum).toEqual(0xfc);

      let unpackedString = "";
      readResponse.payload.sRecords.forEach(
        (srecord) =>
          srecord.sRecordType > FullSRecordMarker.S0_RECORD_TYPE &&
          srecord.sRecordType < FullSRecordMarker.S9_RECORD_TYPE &&
          (unpackedString += srecord.data.toString())
      );

      expect(unpackedString).toContain(inputContent);
    });
  });
});


jest.config.js
/*
 * For a detailed explanation regarding each configuration property and type check, visit:
 * https://jestjs.io/docs/en/configuration.html
 */

module.exports = {
    // All imported modules in your tests should be mocked automatically
    // automock: false,

    // Stop running tests after `n` failures
    // bail: 0,

    // The directory where Jest should store its cached dependency information
    // cacheDirectory: "/private/var/folders/25/jcrpdck169v3v2_zdflhn76m0000gn/T/jest_dx",

    // Automatically clear mock calls and instances between every test
    clearMocks: true,
    // Indicates whether the coverage information should be collected while executing the test
    // collectCoverage: false,

    // An array of glob patterns indicating a set of files for which coverage information should be collected
    // collectCoverageFrom: undefined,

    // The directory where Jest should output its coverage files
    // coverageDirectory: undefined,

    // An array of regexp pattern strings used to skip coverage collection
    // coveragePathIgnorePatterns: [
    //   "/node_modules/"
    // ],

    // Indicates which provider should be used to instrument code for coverage
    // coverageProvider: "babel",

    // A list of reporter names that Jest uses when writing coverage reports
    // coverageReporters: [
    //   "json",
    //   "text",
    //   "lcov",
    //   "clover"
    // ],

    // An object that configures minimum threshold enforcement for coverage results
    // coverageThreshold: undefined,

    // A path to a custom dependency extractor
    // dependencyExtractor: undefined,

    // Make calling deprecated APIs throw helpful error messages
    // errorOnDeprecated: false,

    // Force coverage collection from ignored files using an array of glob patterns
    // forceCoverageMatch: [],

    // A path to a module which exports an async function that is triggered once before all test suites
    // globalSetup: undefined,

    // A path to a module which exports an async function that is triggered once after all test suites
    // globalTeardown: undefined,

    // A set of global variables that need to be available in all test environments
    // globals: {},

    // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
    // maxWorkers: "50%",

    // An array of directory names to be searched recursively up from the requiring module's location
    // moduleDirectories: [
    //   "node_modules"
    // ],

    // An array of file extensions your modules use
    // moduleFileExtensions: [
    //   "js",
    //   "json",
    //   "jsx",
    //   "ts",
    //   "tsx",
    //   "node"
    // ],

    // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
    // moduleNameMapper: {},

    // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
    // modulePathIgnorePatterns: [],

    // Activates notifications for test results
    // notify: false,

    // An enum that specifies notification mode. Requires { notify: true }
    // notifyMode: "failure-change",

    // A preset that is used as a base for Jest's configuration
    // preset: undefined,

    // Run tests from one or more projects
    // projects: undefined,

    // Use this configuration option to add custom reporters to Jest
    // reporters: undefined,

    // Automatically reset mock state between every test
    // resetMocks: false,

    // Reset the module registry before running each individual test
    // resetModules: false,

    // A path to a custom resolver
    // resolver: undefined,

    // Automatically restore mock state between every test
    // restoreMocks: false,

    // The root directory that Jest should scan for tests and modules within
    // rootDir: undefined,

    // A list of paths to directories that Jest should use to search for files in
    roots: [
        "src"
    ],

    // Allows you to use a custom runner instead of Jest's default test runner
    // runner: "jest-runner",

    // The paths to modules that run some code to configure or set up the testing environment before each test
    // setupFiles: [],

    // A list of paths to modules that run some code to configure or set up the testing framework before each test
    // setupFilesAfterEnv: [],

    // The number of seconds after which a test is considered as slow and reported as such in the results.
    // slowTestThreshold: 5,

    // A list of paths to snapshot serializer modules Jest should use for snapshot testing
    // snapshotSerializers: [],

    // The test environment that will be used for testing
    testEnvironment: "node",

    // Options that will be passed to the testEnvironment
    // testEnvironmentOptions: {},

    // Adds a location field to test results
    // testLocationInResults: false,

    // The glob patterns Jest uses to detect test files
    // testMatch: [
    //   "**/__tests__/**/*.[jt]s?(x)",
    //   "**/?(*.)+(spec|test).[tj]s?(x)"
    // ],

    // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
    // testPathIgnorePatterns: [
    //   "/node_modules/"
    // ],

    // The regexp pattern or array of patterns that Jest uses to detect test files
    // testRegex: [],

    // This option allows the use of a custom results processor
    // testResultsProcessor: undefined,

    // This option allows use of a custom test runner
    // testRunner: "jasmine2",

    // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
    // testURL: "http://localhost",

    // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
    // timers: "real",

    // A map from regular expressions to paths to transformers
    // transform: undefined,

    // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
    // transformIgnorePatterns: [
    //   "/node_modules/",
    //   "\\.pnp\\.[^\\/]+$"
    // ],

    // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
    // unmockedModulePathPatterns: undefined,

    // Indicates whether each individual test should be reported during the run
    verbose: true,

    // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
    // watchPathIgnorePatterns: [],

    // Whether to use watchman for file crawling
    // watchman: true,
};


src/utilities/utilities.ts
export const readSubstringAndCastAsNumber = (
  data: string,
  start: number,
  dataLength: number,
  radix: number
): number => parseInt(data.substring(start, start + dataLength), radix);

export const readSubstring = (
  data: string,
  start: number,
  dataLength: number
): string => data.substring(start, start + dataLength);

export const parseHexString = (hex: string): Buffer => {
  const result: number[] = [];
  while (hex.length >= 2) {
    result.push(parseInt(hex.substring(0, 2), 16));
    hex = hex.substring(2, hex.length);
  }

  const buf = Buffer.alloc(result.length);
  result.forEach((b, i) => buf.writeUInt8(b, i));
  return buf;
};

export const decToHex = (dec: number): string => {
  const hex = dec.toString(16).toUpperCase();
  return hex.length % 2 ? hex.padStart(hex.length + 1, "0") : hex;
};


src/types/index.ts
/**
 * General Types
 */
export enum ResponseCodes {
  ERROR,
  WARNING,
  SUCCESS,
}

export enum LogLevel {
  NONE,
  INFO,
  DEBUG,
  ALL,
}

export type StandardisedResponse = {
  success: boolean;
  responseCode: number;
  errors?: Array<Error>;
  message?: string;
  payload: any;
};

export type ReadS19Payload = {
totalFullSRecords: number;
      sRecords: FullSRecordReferenceStructure[];
      totalDataSize: number
}

export interface ReadS19Response extends StandardisedResponse {
  payload: ReadS19Payload;
}


export type S19String = string;

export interface WriteS19Response extends StandardisedResponse {
  payload: S19String;
}


/**
 * S-Record Types
 */

export enum AddressSpaceBoundaries {
  _16_BIT = 65535,
  _24_BIT = 16777215,
  _32_BIT = 4294967295,
}

export type FullSRecordReferenceStructure = {
  sRecordType: FullSRecordMarker;
  data: Buffer;
  dataLength: number;
  address: number;
  checksum: number;
};

export enum FullSRecordMarker {
  START_OF_RECORD = 0x53,
  S0_RECORD_TYPE = 0x30,
  S1_RECORD_TYPE = 0x31,
  S2_RECORD_TYPE = 0x32,
  S3_RECORD_TYPE = 0x33,
  S5_RECORD_TYPE = 0x35,
  S7_RECORD_TYPE = 0x37,
  S9_RECORD_TYPE = 0x39,
}

export enum FullSRecordSizes {
  S0_RECORD_SIZE = 22,
  S1_RECORD_SIZE = 42,
  S2_RECORD_SIZE = 44,
  S3_RECORD_SIZE = 46,
  S5_RECORD_SIZE = 10,
  S7_RECORD_SIZE = 0, // Unknown
  S9_RECORD_SIZE = 10,
}

export enum FullS0ByteLengths {
  START_OF_RECORD_L = 1,
  RECORD_TYPE_L = 1,
  DATA_LENGTH_L = 2,
  ADDRESS_L = 4,
  CHECKSUM_L = 2,
}
export enum FullS1ByteLengths {
  START_OF_RECORD_L = 1,
  RECORD_TYPE_L = 1,
  DATA_LENGTH_L = 2,
  ADDRESS_L = 4,
  CHECKSUM_L = 2,
}
export enum FullS2ByteLengths {
  START_OF_RECORD_L = 1,
  RECORD_TYPE_L = 1,
  DATA_LENGTH_L = 2,
  ADDRESS_L = 6,
  CHECKSUM_L = 2,
}
export enum FullS3ByteLengths {
  START_OF_RECORD_L = 1,
  RECORD_TYPE_L = 1,
  DATA_LENGTH_L = 2,
  ADDRESS_L = 8,
  CHECKSUM_L = 2,
}

export enum FullS5ByteLengths {
  START_OF_RECORD_L = 1,
  RECORD_TYPE_L = 1,
  DATA_LENGTH_L = 2,
  ADDRESS_L = 4,
  CHECKSUM_L = 2,
}

export enum FullS9ByteLengths {
  START_OF_RECORD_L = 1,
  RECORD_TYPE_L = 1,
  DATA_LENGTH_L = 2,
  ADDRESS_L = 4,
  CHECKSUM_L = 2,
}
